sdk/dotnet/Docker.cs

```diff
@@ -0,0 +1,699 @@
+
+            var stderr = new StringBuilder();
+            process.StartInfo.RedirectStandardError = true;
+            process.ErrorDataReceived += new DataReceivedEventHandler((object sender, DataReceivedEventArgs e) =>
+            {
+                // We can't stream these stderr messages as we receive them because we don't knows at
+                // this point because Docker uses stderr for both errors and warnings.  So, instead, we
+                // just collect the messages, and wait for the process to end to decide how to report them.
+                stderr.Append(e.Data ?? "");
+            });
+
+            try
+            {
+                process.Start();
+                process.BeginOutputReadLine();
+                process.BeginErrorReadLine();
+
+                if (stdin != null)
+                {
+                    process.StandardInput.Write(stdin);
+                    process.StandardInput.Close();
+                }
+
+                await WaitForExitAsync(process);
+                var code = process.ExitCode;
+
+                // If we got any stderr messages, report them as an error/warning depending on the
+                // result of the operation.
+                if (stderr.Length > 0)
+                {
+                    if (code > 0 && !reportErrorAsWarning)
+                    {
+                        // Command returned non-zero code.  Treat these stderr messages as an error.
+                        Log.Error(stderr.ToString(), logResource);
+                    }
+                    else
+                    {
+                        // command succeeded.  These were just warning.
+                        Log.Warn(stderr.ToString(), logResource);
+                    }
+                }
+
+                // If the command failed report an ephemeral message indicating which command it was.
+                // That way the user can immediately see something went wrong in the info bar.  The
+                // caller (normally runCommandThatMustSucceed) can choose to also report this
+                // non-ephemerally.
+                if (code > 0)
+                {
+                    Log.Info(GetFailureMessage(cmd, args, reportFullCommandLine, code), logResource, ephemeral: true);
+                }
+
+                return (code, stdout.ToString());
+            }
+            catch (Exception ex)
+            {
+                // This shouldn't normally happen, but we want to be sure Process doesn't throw.
+                // In case it does, return an error with the exception printout.
+                return (1, ex.ToString());
+            }
+        }
+
+        private static async Task WaitForExitAsync(Process process, CancellationToken cancellationToken = default)
+        {
+            var tcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            void ProcessExited(object? sender, EventArgs e) => tcs.TrySetResult(true);
```

Comment:
can you place local functions att the end of the containing method.  also, place a `return;` statement above it.  it makes it very clear waht the separation is between teh normal code and helper functions.

Reply:
done